gql/dsl.py

```diff
@@ -1,179 +1,362 @@
+import logging
 from collections.abc import Iterable
-from functools import partial
+from typing import Any, Callable, Dict, List, Optional, Union, cast
 
 from graphql import (
     ArgumentNode,
     DocumentNode,
     EnumValueNode,
     FieldNode,
     GraphQLEnumType,
+    GraphQLField,
     GraphQLInputField,
     GraphQLInputObjectType,
+    GraphQLInputType,
+    GraphQLInterfaceType,
     GraphQLList,
+    GraphQLNamedType,
     GraphQLNonNull,
+    GraphQLObjectType,
+    GraphQLScalarType,
+    GraphQLSchema,
     ListValueNode,
     NameNode,
     ObjectFieldNode,
     ObjectValueNode,
     OperationDefinitionNode,
     OperationType,
     SelectionSetNode,
+    ValueNode,
     ast_from_value,
     print_ast,
 )
 from graphql.pyutils import FrozenList
 
 from .utils import to_camel_case
 
+log = logging.getLogger(__name__)
 
-class DSLSchema(object):
-    def __init__(self, client):
-        self.client = client
+GraphQLTypeWithFields = Union[GraphQLObjectType, GraphQLInterfaceType]
+Serializer = Callable[[Any], Optional[ValueNode]]
 
-    @property
-    def schema(self):
-        return self.client.schema
 
-    def __getattr__(self, name):
-        type_def = self.schema.get_type(name)
+def dsl_gql(*fields: "DSLField") -> DocumentNode:
+    """Given arguments of type :class:`DSLField` containing GraphQL requests,
+    generate a Document which can be executed later in a
+    gql client or a gql session.
+
+    Similar to the :func:`gql.gql` function but instead of parsing a python
+    string to describe the request, we are using requests which have been generated
+    dynamically using instances of :class:`DSLField` which have been generated
+    by instances of :class:`DSLType` which themselves have been generated from
+    a :class:`DSLSchema` class.
+
+    The fields arguments should be fields of root GraphQL types
+    (Query, Mutation or Subscription).
+
+    They should all have the same root type
+    (you can't mix queries with mutations for example).
+
+    :param fields: root instances of the dynamically generated requests
+    :type fields: DSLField
+    :return: a Document which can be later executed or subscribed by a
+        :class:`Client <gql.client.Client>`, by an
+        :class:`async session <gql.client.AsyncClientSession>` or by a
+        :class:`sync session <gql.client.SyncClientSession>`
+
+    :raises TypeError: if an argument is not an instance of :class:`DSLField`
+    :raises AssertionError: if an argument is not a field of a root type
+    """
+
+    # Check that we receive only arguments of type DSLField
+    # And that they are a root type
+    for field in fields:
+        if not isinstance(field, DSLField):
+            raise TypeError(
+                f"fields must be instances of DSLField. Received type: {type(field)}"
+            )
+        assert field.type_name in ["Query", "Mutation", "Subscription"], (
```

Comment:
Shouldn't this assertion match the enum types of the spec for [OperationType](https://spec.graphql.org/draft/#OperationType)?

Also graphql-core has [OperationType](https://github.com/graphql-python/graphql-core/blob/df8b5484a9e3ed69b1dcad66e923f9d1fea2cbdc/src/graphql/language/ast.py#L210) as an Enum.